### DISCUSSION TITLE
How z3 deal with define-fun?

### DISCUSSION THREAD

**Discussion by zxt5** [Z3Prover/z3#6445] [Q&A] [answered]
From the answer of this question ![Difference between macro and quantifier in Z3](https://stackoverflow.com/questions/15758699/difference-between-macro-and-quantifier-in-z3), z3 treats the function definition as a macro, and replaces every occurrence of the function with its expression. But the answer is from 9 years ago, so I'm wondering it there any differences from the current z3? How z3 deal with `define-fun` now?

Thanks!

**Comment by NikolajBjorner** [comment, marked as answer]
Use define-fun-rec if you want to use use the function declaration inside of an array map. Example, 

```
(define-fun-rec f ((x Int)) Int (+ x 1))

(declare-const A (Array Int Int))
(declare-const B (Array Int Int))

(assert (= ((_ map f) A) B))
(check-sat)
(get-model)
```

define-fun does is (still) treated as a macro with eager unfolding (= without lazy unfolding).

**Reply by zxt5** [reply]
Thank you!
